### ISSUE TITLE
Update RxJS in user test suite

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#33125]
Sounds like something slipped by us this release.

https://twitter.com/BenLesh/status/1166838448688381953

**Comment by DanielRosenwasser** [comment]
@benlesh got any details you can share on what went wrong?

**Comment by benlesh** [comment]
I'm currently trying to update master to 3.6.2, and just running into interesting new typing issues. Probably a dozen or so in total. Most of them are probably a correctness thing, I'm sure.

You can try it though:

1. Pull RxJS from this sha: https://github.com/ReactiveX/rxjs/commit/c311d5b1a50527000c3a193516a6876d8be9543d
2. `npm i -D typescript@3.6`
3. `npm run build_all`

:)

I'm not complaining, but I swear that 90% of RxJS work is just updating typings and dealing with the fallout of type updates, new inference patterns, and changes to strictness. :) It's just the job now, I guess. haha

**Comment by benlesh** [comment]
Some of it seems to stem from some interesting trickery we're doing to type things like `switchMap`, such that they're able to properly infer types from union returns.

Prior to doing some trickery with conditional types, this didn't work:

```ts
import { of } from 'rxjs';
import { switchMap } from 'rxjs';

of(Math.random()).pipe(switchMap(n => n > 0.5 ? of('test') : of(123))) // $ExpectType Observable<string | number>
```

**Comment by weswigham** [comment]
I think the Rx in our user suite is just the published `npm` package, not their input source build. We'd probably need to add it properly as a built item.

**Comment by benlesh** [comment]
Yeah, there's some incompatibility with `@types/node` it looks like:

`node_modules/@types/node/index.d.ts:74:11 - error TS2300: Duplicate identifier 'IteratorResult'.

74 interface IteratorResult<T> { }
             ~~~~~~~~~~~~~~

  node_modules/typescript/lib/lib.es2015.iterable.d.ts:41:6
    41 type IteratorResult<T, TReturn = any> = IteratorYieldResult<T> | IteratorReturnResult<TReturn>;
            ~~~~~~~~~~~~~~
    'IteratorResult' was also declared here.

node_modules/typescript/lib/lib.es2015.iterable.d.ts:41:6 - error TS2300: Duplicate identifier 'IteratorResult'.

41 type IteratorResult<T, TReturn = any> = IteratorYieldResult<T> | IteratorReturnResult<TReturn>;
        ~~~~~~~~~~~~~~

  node_modules/@types/node/index.d.ts:74:11
    74 interface IteratorResult<T> { }
                 ~~~~~~~~~~~~~~
    'IteratorResult' was also declared here.``
```

**Comment by OliverJAsh** [comment]
We just tried to upgrade unsplash.com from TS 3.5.2 to 3.6.2 and ran into 2 regressions, which I've filed:

- #33132
- #33133

I'd really love for Unsplash to be part of the TS user test suite as well.

**Comment by felixfbecker** [comment]
Related: #33131

**Comment by benlesh** [comment]
It looks like the biggest problems right now are failures around a solution we were using to get types to infer properly from some of our functions. [Here you can find a minimal reproduction of what we were doing](http://www.typescriptlang.org/play/#code/PQKhAIAUCcHsCMA2BTAtuEwBQWBmBXAOwGMAXAS1kPFNgElTloBDW6AHgBUA+ACnMYskyAFzgGTZsK7cAlGIks2M8AG8s4TeGjJS+aNQGThAbQDKAT1TxYiAHRGlsaAF1esgNxYAvjizEqAGdScGZwAF4aekFWZ14AckZg+M9-IJD4CKjFWOheEwBGABpwACYSgGYXVIDCYPBiLNoctl4AWVYACzsWQgATWFR3cAB+cETkZPAxQpLy8CrUnFAIM1t8CioMbCxSCwAHZHEYxj6ANWZEfGQAeVx2Om4sunBkAA9GfsDj4xR2ckIuCY4B4oxB03AhGQADcmF48EQyJRqM0Ts4AGLkN7IPpcV4fZBfH5CP7MQgWbh8RxSFBiTjyYm5B4nHEXK63e48J7qLTaXT6QwxUyWay2Bxo1zuLy+NJ1EJ9JrRSRsTHYvoJJKkFJeWr1I6RVHKjFYnH5YplSrVHXpcC4RUtY1q9pdHpkgZDWRgiZTGbm+aLDxAA)

Basically this:

```ts
/** Problem */

function toIterator<T>(iterable: Iterable<T>): Iterator<T> {
    return iterable[Symbol.iterator]();
}


const a = toIterator('test');
const b = toIterator([1, 2, 3]);
const c = toIterator(Math.random() ? 'test' : [1, 2, 3]); // ERROR HERE

/** Solution */

type IteratedValueOf<I> = I extends Iterable<infer T> ? T : never;

function toIteratorFixed<T extends Iterable<any>>(iterable: T): Iterator<IteratedValueOf<T>> {
    return iterable[Symbol.iterator]();
}

const d = toIteratorFixed('test');
const e = toIteratorFixed([1, 2, 3]);
const f = toIteratorFixed(Math.random() ? 'test' : [1, 2, 3]); // FIXED HERE
```

**Comment by weswigham** [comment]
cc @ahejlsberg the above example is connected with the union type matching behavior we added into inference.

**Comment by weswigham** [comment]
Although probably just a dupe of #33131 at this point but I'll link the two issues anyway for context.
